storage/src/main/java/org/apache/uniffle/storage/handler/impl/DataSkippableReadHandler.java

```diff
@@ -67,8 +68,17 @@ public DataSkippableReadHandler(
 
   public ShuffleDataResult readShuffleData() {
     if (shuffleDataSegments.isEmpty()) {
-      ShuffleIndexResult shuffleIndexResult = readShuffleIndex();
+      ShuffleIndexResult shuffleIndexResult;
+      try {
+        shuffleIndexResult = readShuffleIndex();
+      } catch (Exception e) {
+        if (++failTimes > maxHandlerFailTimes) {
+          isFinished = true;
+        }
+        throw e;
```

Comment:
I don't understand here
```
        if (++failTimes > maxHandlerFailTimes) {
          isFinished = true;
        }
        throw e;
```
If we throw a exception to upper layer,  I feel that it means that the handler is finished. Why do we need to reach the max failure time and then set the `isFinished` value to the `true`?
Could we add a test case for this situation to help us understand?

Reply:
1.If this exception thrown here, the upper handler will catch it and move to next lower handler.
2.if the network temporarily unavailable, read shuffle index may fail here,so i think we should retry in next round, until reach `maxHandlerFailTimes`.

Reply:
For 2nd point:
If the network is unavailable temporarily, we won't retry this handler because we throw the exception here. Is it necessary that we add the failTimes?

Reply:
> we won't retry this handler because we throw the exception here

Why not? It just will not retry in current round, but it will retry in next round.

Reply:
> > we won't retry this handler because we throw the exception here
> 
> Why not? It just will not retry in current round, but it will retry in next round.

If upper layer is LocalFileQuorumClientReadHanlder, we will retry another handler when we throw an exception.
https://github.com/apache/incubator-uniffle/blob/4a3d2be36795df2eaf2edbe75ecc5816bf7eb87a/storage/src/main/java/org/apache/uniffle/storage/handler/impl/LocalFileQuorumClientReadHandler.java#L105

It seems that fail handler times won't reach the maxHandlerFailTimes  in any time.   Is  it necessary that we increase the failTimes? Could we change the code?
```
        if (++failTimes > maxHandlerFailTimes) {
          isFinished = true;
        }
        throw e
```
to 
```
isFinished = true;
throw e
```
Could you add a test case to explain when the fail time will reach the maxHandlerFailTimes?

Reply:
> Could you add a test case to explain when the fail time will reach the maxHandlerFailTimes?

Done

Reply:
I see your case. Terrific! But in my understanding the fail times of LocalFileQuorumClientReadHandler reach the  maxHandlerFailTimes, LocalFileQuorumClientReadHandler have three LocalFileClientReadHandlers. Every  LocalFileClientReadHandler don't reach maxHandlerFailTimes. I mean that the fail times of LocalFileClientReadHandler  isn't meaningful. So could we change the code like my sayings? or the fail times of DataSkippableReadHandler is only used for HDFSFileClientReadHandler? 
My concern is that the fail times of LocalFileClientReadHandler seems meaningless.

Reply:
> But in this case the fail times of LocalFileQuorumClientReadHandler reach the maxHandlerFailTimes, LocalFileQuorumClientReadHandler have three LocalFileClientReadHandlers. Every LocalFileClientReadHandler don't reach maxHandlerFailTimes

`LocalFileQuorumClientReadHandler`,  `MemoryQuorumClientReadHandler` and `HdfsClientReadHandler` have no `maxHandlerFailTimes`. If all LocalFileClientReadHandlers finished, it is finished.

Reply:
This means a little complex. It's better that we can unify the behavior of handlers. We hope that we have a universe framework for handlers.Is it possible to accomplish this aim?

Reply:
The problem which we want to solve is that we want to read correct data when one replica don't have complete data. Should we only modify the quorum handler's logic? We don't expose the logic to upper layer. We process the fault tolerance in the quorum handler.

Reply:
> > Should we only modify the quorum handler's logic?
> 
> No, i think we need retry in ShuffleClient, we don't know whether we have read complete data until read from all quorum handlers in each round, and we need to check blocks consistent each round.

If we pass the expectBlockId to the quorum handler, could we check blocks in the quorum handler?